keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -275,20 +275,48 @@ with a metric.
+
+We need to watch out for inadvertantly putting apiserver IP/endpoint information and the trust bundle control (used to authenticate server while proxying) in user hands (by exposing this information in REST APIs)
 
 ## Design Details
 
-TODO: explanation of how the handler will determine a request is for a resource
-that should be proxied.
+### Aggregation Layer
+
+1. A new filter will be added to the [handler chain] of the aggregation layer. This filter will maintain an internal map with the key being the group-version-resource and the value being a list of server IDs of apiservers that are capable of serving that group-version-resource.
+
+    1. This internal map is populated using an informer for StorageVersion objects. An event handler will be added for this informer that will get the apiserver ID of the requested group-version-resource and update the internal map accordingly.
+
+2. This filter will pass on the request to the next handler in the local aggregator chain, if:
+    1. It is a non resource request
+    2. No StorageVersion was retrieved for it, meaning the request is for an aggregated API or for a custom resource
+    3. If the local apiserver ID is found in the list of serviceable-by server IDs from the internal map
+
+3. If the local apiserver ID is not found in the list of serviceable-by server IDs, a random apiserver ID will be selected from the retrieved list and the request will be proxied to this apiserver.
+
+4. If there is no apiserver ID retrieved for the requested GVR, we will serve 503 with error `No relevant API server found for the requested GVR <group_version_resource>`
+
+[handler chain]:https://github.com/kubernetes/kubernetes/blob/fc8f5a64106c30c50ee2bbcd1d35e6cd05f63b00/staging/src/k8s.io/apiserver/pkg/server/config.go#L639
+
+In order to ensure that we do not end up rerouting the request indefinitely from one apiserver to another, we will ensure that we short circuit an already proxied request to be served locally, by adding a header to the proxied request. Then, when the destination apiserver finds this header in the request, we should not proxy the request again.
+
+#### Identifying destination apiserver's network location
+
+* TODO: We need to find a place to store and retrieve the destination apiserver's host and port information given the server's ID.
+We do not want to store this information in
 
-TODO: explanation of how the security handshake between apiservers works.
-* What we need to fix: random processes / external users / etc should not be
-  able to proxy requests, so the receiving apiserver needs to be able to verify
-  the source apiserver.
-* generate self-signed cert on startup, put pubkey in apiserver identity lease
-  object?
+    * StorageVersion : because we do not want to expose the network identity of the apiservers in this API that can be listed in places where it could be unsafe to do so
```

Comment:
The problem is it would be highly redundant, it's not about safety

Reply:
Corrected, thanks!